enhancements/sig-architecture/32-extensiblescheduling/score-generator/README.md

```diff
@@ -0,0 +1,192 @@
+status:
+  conditions:
+  - lastTransitionTime: "2021-10-28T08:31:39Z"
+    message: AddOnPlacementScore updated successfully
+    reason: AddOnPlacementScoreUpdated
+    status: "True"
+    type: AddOnPlacementScoreUpdated
+  validUntil: "2021-10-29T18:31:39Z"
+  scores:
+  - name: "redhat-com-avgLatency"
+    value: 30
+  - name: "google-com-avgLatency"
+    value: 50
+~~~
+
+Now, a `Placement` like this could be used:
+
+~~~yaml
+apiVersion: cluster.open-cluster-management.io/v1beta1
+kind: Placement
+metadata:
+  name: latency-placement
+  namespace: ns1
+spec:
+  numberOfClusters: 3
+  prioritizerPolicy:
+    mode: Exact
+    configurations:
+      - scoreCoordinate:
+          type: AddOn
+          addOn:
+            resourceName: cluster1-generator
+            scoreName: redhat-com-avgLatency
+        weight: -1
+~~~
+
+Other applications may use latency to google.com.
+
+Now, we want to add linux.com to the list of locations to test. With the current implementation we 
+will need to edit the code of the addon controller and include the test to linux.com + the result to be added
+to the `AddOnPlacementScore`.
+
+To fix above issue, the proposal is to create a new API `AddOnPlacementScoreGenerator`, which could 
+look like this for the example above:
+
+~~~yaml
+apiVersion: cluster.open-cluster-management.io/v1alpha1
+kind: AddOnPlacementScoreGenerator
+metadata:
+  name: cluster1-generator
+  namespace: cluster1
+spec:
+  addOnPlacementSelector:
+    name: latency
+    namespace: cluster1
+  latencies:
+    - name: redhat-com-avgLatency
+      url: https://redhat.com
+      runs: 2
+      waitBetweenRuns: 10s
+    - name: google-com-avgLatency
+      url: https://google.com
+    - name: linux-com-avgLatency
+      url: https://linux.com
+~~~
+
+Our controller can now read the `AddOnPlacementScoreGenerator` and our specific controller will be interested on 
+everything below `.spec.latencies`. For example, the redhat-com-avgLatency will have the result of running two latency 
+tests to https://redhat.com and waiting 10s between runs, then the mean value will be posted to the `AddOnPlacementScore` 
+status.
+
+## Goals & Non-Goals
+
+### Goals
+
+- Influence the behavior of addon controllers updating `AddOnPlacementScores`
+
+### Future goals
+
+- It is currently assumed that the user of `AddOnPlacementScoreGenerator` is either a 
+cluster admin or a user who can create `AddOnPlacementScoreGenerator` in the hub cluster's 
+managed "cluster namespace".
+
+## Design
+
+### Component & API
+
+We purpose to adding a new custom resource named
+`AddOnPlacementScoreGenerator` introduced into OCM by this proposal:
+
+A sample of the `AddOnPlacementScoreGenerator` will be:
+
+~~~yaml
+apiVersion: cluster.open-cluster-management.io/v1alpha1
+kind: AddOnPlacementScoreGenerator
+metadata:
+  name: cluster1-generator
+  namespace: cluster1
+spec:
+  addOnPlacementSelector:
+    name: latency
+    namespace: cluster1
+  latencies:
```

Comment:
My question is, should this field `latencies` be a more general field? and how could a user define a URL here, what's the requirement for the URL and the output?

Reply:
Hey @haoqing0110 yes, other than more generic to me this would be "any string" and even we could have more than one, something like this for example: 

~~~yaml
apiVersion: cluster.open-cluster-management.io/v1alpha1
kind: AddOnPlacementScoreGenerator
metadata:
  name: cluster1-generator
  namespace: cluster1
spec:
  addOnPlacementSelector:
    name: latency
    namespace: cluster1
  latencies:
    - name: redhat-com-avgLatency
      url: https://redhat.com
      runs: 2
      waitBetweenRuns: 10s
    - name: google-com-avgLatency
      url: https://google.com
    - name: linux-com-avgLatency
      url: https://linux.com
  energy:
    - name: some-name
      someValue: some-value
~~~

Reply:
Thanks for the explanation. This `AddOnPlacementScoreGenerator` sounds like a configuration for the Addon controller, and to generate different scores, developers still need to develop their own addon controller, correct?  Sounds similar to the idea of https://github.com/open-cluster-management-io/enhancements/tree/main/enhancements/sig-architecture/58-addon-configuration, not sure if this can satisfy the motivation of this proposal.

Reply:
cc @qiujian16 , want to know your suggestions.

Reply:
@haoqing0110 the use cases defined by 58-addon-configuration would work for us, I have checked the examples and the example with `ManagedProxyConfiguration`  is very similar to what we want.

And answering your second comment, yes, the same controller could use `latencies` and `energy` data.

Reply:
I think this is more like a cluster scoped resource that applies to all clusters, it is less likely that we want different configs for different clusters. Also how could we parse the result of the http responses? It might be straightforward for latency. But if we try to fetch data from a metrics server or 3rd party services to generate the score, the returned output varies and we might need a way to define how to parse the result.

Reply:
Hey @qiujian16,

I think that for latencies global object or per-cluster object would both work. In terms of data,  I was thinking of an unstructured structure where people developing custom controllers with the addon framework will consume the data the way they need it. I don't think this should be a controller taking care of multiple data sources, instead a single controller taking care of a data source, so if anything were wrong with the data itself it wouldn't affect the other add-ons.

Reply:
+1 on Mario's approach on a single controller taking care of a single data source type.